### DISCUSSION TITLE
High Availability Setup for Postal Across Regions

### DISCUSSION THREAD

**Discussion by emahdij** [postalserver/postal#3182] [Help with using Postal]
Hi everyone,

I've set up Postal with 10 IP addresses and IP pools, and I'm looking to make it highly available. I read some discussions here about HA setups and saw suggestions for using HA MySQL clusters. However, since I plan to deploy Postal in two different regions with high latency between them, setting up a shared MySQL cluster isn’t feasible.

My emails are sent through Postal's API, and my application can manage Postal's health checks. I'm thinking of running two separate Postal instances under the same domain, using the same DNS records (like DKIM settings), without adding extra records.

Is this approach advisable? Could it affect my domain or IP reputation? Also, if I restore the MySQL data to the second server, what challenges might arise?

Has anyone tried a similar setup or encountered any issues? Any advice would be appreciated!

**Comment by willpower232** [comment]
Postal is fully clusterable so you should be able to run as many copies as you want, as long as they're all pointing to the same pile of databases then you shouldn't have any issues from that perspective.

What is more important than your databases is ensuring that all your IP addresses and pools work on all instances that you are running.

As Postal is the one sending your messages, you also need to make sure that all its IP addresses are mentioned in the SPF records one way or the other. The DKIM records should be fine but it would also be worth ensuring that you have reverse DNS set up on the IP addresses as well.

Regarding restoring database data, you will need to be aware that the queue is fully database managed so if your copy of the data is slightly out of sync then some messages could be delivered twice or missing from the data altogether.

**Reply by emahdij** [reply]
Thanks for the reply.

The main issue is that the databases won't be in sync. As I mentioned earlier, my application will call the Postal API and decide which Postal instance to send through, depending on the zone. The databases are separate for each instance, but I'll restore the data from the first database to the second server to keep the DKIM and other records consistent. I've also added PTR records and updated the SPF.

Is there anything about this setup that could cause problems? Any structural issues I should be aware of?

**Reply by willpower232** [reply]
If the same domains are in multiple places then you will have to look in multiple places to see what happened to them.

Postal would fail to label a message as bounced if one Postal sent the message and the other received the bounce. You would probably want to fully set up webhooks so you could try and figure out the bounces yourself.

If you're sending messages for the domain from both places then the suppression list could be annoying if one server was blocking a recipient and the other wasn't. You'd also have to check both places to unblock someone.

**Reply by emahdij** [reply]
Thanks, I hadn’t thought about that. I’ll make sure to handle this issue from my service. Aside from managing bounces, are there any other potential issues or considerations I should be aware of?

**Reply by willpower232** [reply]
I think I've mentioned everything I can think of. Using a database cluster with a reader copy and automatic failover would probably simplify the situation somewhat but definitely not a cheap option.

**Reply by emahdij** [reply]
I've set up the second server. I tried restoring the data, but it didn't work, and the API returned a 5xx error. I added the domain in the second console, and now it's working.

The last thing I need help with is the DKIM record. I’d prefer to have a single DKIM record, but currently, they are different. I searched the discussions but didn’t find a solution, except for some suggestions about making changes in the database. Could you help me with how to update the DKIM?

**Reply by willpower232** [reply]
If you have manually added the domains to both copies of Postal, you'll have to rummage in the database to find what it thinks the DKIM information is and make it the same between all servers which use the domain. It sounds like this information wasn't in the data copy you tried to transfer across.
